lib/web_ui/lib/src/engine.dart

```diff
@@ -203,6 +203,7 @@ void initializeEngine() {
   // This extension does not need to clean-up Dart statics. Those are cleaned
   // up by the compiler.
   developer.registerExtension('ext.flutter.disassemble', (_, __) {
+    window.resetHistory();
```

Comment:
The pattern that we use in the web engine is to call `registerHotRestartListener` and pass a callback. The callback will be invoked when hot restart is requested.

So somewhere near the code that initializes the history, you could do:
```dart
registerHotRestartListener(() {
  window.resetHistory();
});
```

Example: https://github.com/flutter/engine/blob/42019c8361572ec684e4d2a4d7cc5141394f5855/lib/web_ui/lib/src/engine/platform_dispatcher.dart#L768-L770